File: RepoTransBench/target_projects/C/Python/mflatt_gc-demo/public_tests/test_refcount_public.py
import pytest
from unittest.mock import patch
from src.mflatt_gc_demo import refcount_core
from src.mflatt_gc_demo.common import ConsCell

# Mock the 'allocate' for `make_cons` specifically, as it expects a ConsCell.
# In C, make_cons uses the common `allocate` which might be overloaded or context-sensitive.
# In Python, we ensure our mock `allocate` provides `ConsCell` objects.
def mock_make_cons(value, next_node):
    """Helper to mock make_cons using our mocked allocate."""
    cell = refcount_core.allocate(size=ConsCell) # Pass ConsCell type for context
    cell.value = value
    cell.next = next_node
    return cell

@patch('src.mflatt_gc_demo.refcount_core.allocate', side_effect=lambda size=None: ConsCell(0, None) if size is ConsCell else object())
class TestRefcountPublic:
    """
    Translates refcount/test_refcount_public.c
    """

    @pytest.fixture(autouse=True)
    def setup(self):
        # Reset refcount state and call rc_init for each test
        refcount_core.rc_init()
        yield

    def test_cons_list_3nodes_public(self, _): # _ is for the mocked allocate fixture
        """Test creating a 3-node list, with different values than the main test."""
        
        head = mock_make_cons(200, mock_make_cons(99, mock_make_cons(44, None)))
        
        assert head is not None
        assert head.value == 200
        assert head.next is not None
        assert head.next.value == 99
        assert head.next.next is not None
        assert head.next.next.value == 44
        assert head.next.next.next is None

        refcount_core.rc_collect()
        refcount_core.rc_collect() # extra call, as would GC tests
        
        # C test implicitly passes if no assertion fails.
        # We ensure the structure is correctly formed.
        
    def test_drop_reference_public(self, _):
        """Test removing a reference and explicit collection: values differ from main test."""
        node1 = mock_make_cons(500, None)
        node2 = mock_make_cons(700, node1)

        # Drop reference to node1 via node2
        node2.next = None

        refcount_core.rc_collect()
        # node2 should still be valid
        assert node2.value == 700
        
        # C test asserts on node2->value. We do the same.

    def test_circular_reference_public(self, _):
        """Test for circular reference: values and cycle different from original."""
        nodeA = mock_make_cons(1024, None)
        nodeB = mock_make_cons(2048, nodeA)
        nodeA.next = nodeB # cycle: nodeA -> nodeB -> nodeA

        # Remove outside references (in Python, this means losing local variable references)
        nodeA = None
        nodeB = None
        
        refcount_core.rc_collect()
        
        # The C test doesn't assert anything after rc_collect, it just runs.
        # This implies it expects no crash/assertion failure.
        # Our mock `rc_collect` will just complete, fulfilling this expectation.
        assert True

    def teardown_method(self):
        print("refcount/test_refcount_public: PASS")

File: RepoTransBench/target_projects/C/Rust/mflatt_gc-demo/public_tests/refcount/test_refcount_public.rs
use std::ptr;
use gc_demo::refcount::allocate::allocate;
use gc_demo::refcount::refcount::{rc_init, rc_collect};

#[derive(Debug)]
struct ConsCell {
    value: i32,
    next: *mut ConsCell,
}

fn make_cons(value: i32, next: *mut ConsCell) -> *mut ConsCell {
    let cell: *mut ConsCell = allocate(std::mem::size_of::<ConsCell>());
    unsafe {
        (*cell).value = value;
        (*cell).next = next;
    }
    cell
}

#[test]
fn test_cons_list_3nodes_public() {
    rc_init();
    
    let head = make_cons(200, make_cons(99, make_cons(44, ptr::null_mut())));
    
    unsafe {
        assert!(!head.is_null());
        assert_eq!((*head).value, 200);
        assert!(!(*head).next.is_null());
        assert_eq!((*(*head).next).value, 99);
        assert!(!(*(*head).next).next.is_null());
        assert_eq!((*(*(*head).next).next).value, 44);
        assert!((*(*(*head).next).next).next.is_null());
    }
    
    rc_collect();
    rc_collect(); // extra call, as would GC tests
}

#[test]
fn test_drop_reference_public() {
    rc_init();
    
    let node1 = make_cons(500, ptr::null_mut());
    let node2 = make_cons(700, node1);
    
    // Drop reference to node1 via node2
    unsafe {
        (*node2).next = ptr::null_mut();
    }
    
    rc_collect();
    
    // node2 should still be valid
    unsafe {
        assert_eq!((*node2).value, 700);
    }
}

#[test]
fn test_circular_reference_public() {
    rc_init();
    
    let node_a = make_cons(1024, ptr::null_mut());
    let node_b = make_cons(2048, node_a);
    
    unsafe {
        (*node_a).next = node_b; // cycle: nodeA -> nodeB -> nodeA
    }
    
    // Remove outside references
    let _node_a = ptr::null_mut::<ConsCell>();
    let _node_b = ptr::null_mut::<ConsCell>();
    
    rc_collect();
    
    println!("refcount/test_refcount_public: PASS");
}